modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationTombstonesWithIndicesTest.java

```diff
@@ -0,0 +1,236 @@
+        stopAllGrids();
+
+        super.afterTest();
+    }
+
+    /**
+     *
+     */
+    @Parameterized.Parameters(name = "atomicity = {0}, repair = {1}")
+    public static List<Object[]> parameters() {
+        ArrayList<Object[]> params = new ArrayList<>();
+
+        CacheAtomicityMode[] atomicityModes = new CacheAtomicityMode[] {ATOMIC, TRANSACTIONAL};
+
+        for (CacheAtomicityMode atomicityMode : atomicityModes) {
+            params.add(new Object[] {atomicityMode, false});
+            params.add(new Object[] {atomicityMode, true});
+        }
+
+        return params;
+    }
+
+    @Test
+    @WithSystemProperty(key = DEFAULT_TOMBSTONE_TTL_PROP, value = "300000000")
+    public void testTombstones() throws Exception {
+        IgniteEx ig = startGrids(NODES_CNT);
+
+        IgniteEx client = startClientGrid(NODES_CNT);
+
+        ig.cluster().state(ACTIVE);
+
+        Class<?> customKeyCls = ig.configuration().getClassLoader().loadClass(CUSTOM_KEY_CLS);
+        Class<?> customValCls = ig.configuration().getClassLoader().loadClass(CUSTOM_VAL_CLS);
+
+        Constructor<?> keyCtor = customKeyCls.getDeclaredConstructor(int.class);
+
+        IgniteCache<Object, Object> clientCache = client.cache(DEFAULT_CACHE_NAME);
+
+        Object primaryKey = keyCtor.newInstance(42);
+        Object primaryVal = customValCls.newInstance();
+
+        int primaryNodeIdx = -1;
+        for (int i = 0; i < NODES_CNT; ++i) {
+            if (grid(i).affinity(DEFAULT_CACHE_NAME).isPrimary(grid(i).localNode(), primaryKey)) {
+                primaryNodeIdx = i;
+
+                break;
+            }
+        }
+
+        assertTrue("Failed to find primary node for key [key=" + primaryKey + ']', primaryNodeIdx >= 0);
+
+        clientCache.put(primaryKey, primaryVal);
+
+        // Block first NODES_CNT - 1 messages from primary node,
+        // to emulate the case when primary contains a tombstone for the key
+        // and backup nodes store the original value.
+        AtomicInteger blocked = new AtomicInteger(NODES_CNT - 1);
+        TestRecordingCommunicationSpi spi = TestRecordingCommunicationSpi.spi(grid(primaryNodeIdx));
+        spi.blockMessages((node, msg) -> {
+            boolean blockedMsg = cacheAtomicityMode == ATOMIC ?
+                msg instanceof GridDhtAtomicSingleUpdateRequest :
+                msg instanceof GridDhtTxFinishRequest;
+
+            return blockedMsg && blocked.decrementAndGet() >= 0;
+        });
+
+        final int pNodeIdx = primaryNodeIdx;
+        IgniteInternalFuture<?> removeFut = runAsync(() -> {
+            grid(pNodeIdx).cache(DEFAULT_CACHE_NAME).remove(primaryKey);
+        });
+
+        assertTrue(
+            "Failed to wait all update messages that are sent to backup nodes.",
+            waitForCondition(() -> blocked.get() == 0, 10_000));
+
+        ReconciliationResult res = partitionReconciliation(ig, fixMode, RepairAlgorithm.PRIMARY, 4, DEFAULT_CACHE_NAME);
```

Comment:
Am I right that we test only the "primary" algo in this new test?

Reply:
Nope, we have an additional test to check all modes except "print only" https://github.com/gridgain/gridgain/pull/3363/files#diff-3c822b61a4def3cf7e92f94a48019d827dc4c6b42282f2ebfc1f33bfec92a626R466